### PR TITLE
Generate DC test vectors and add other signature schemes

### PR DESCRIPTION
**Pull request by claucece**
Open for comments ;)

cc./ @cjpatton @chris-wood @xvzcf

### CODE REVIEW

**Comment by xvzcf** [comment]
To me, it seems like `-make-dcvectors` and `-make-dc` are the same thing. Also, instead of a `-dc-algo` argument, we could have an `-alg` argument that is passed to `make-dc`, `make-root`, etc. I say `alg` and not `sig-alg` because we might potentially wrap post-quantum KEMs in DCs.

File: cmd/util/common.go [hunk 1]
```diff
@@ -104,6 +104,7 @@ type Config struct {
 	// implementations.
 	Bugs CertificateBugs
 
+	// The signature algorithm for the Delegated Credential
```
Comment (cjpatton) [inline cmd/util/common.go:107]: Doc strings for parameters being with the name of the parameter.
```
// SignatureAlgorithm ....
```

Also, since this variable is exported, I'm wondering if we should export its type as well?

File: Makefile [hunk 2]
```diff
@@ -27,9 +28,20 @@ testinputs: util
...
 
+dcvectors: util
+	mkdir -p ${DCVECTORS_DIR}
+	${UTIL} -make-root -out ${DCVECTORS_DIR}/root.crt -key-out ${DCVECTORS_DIR}/root.key -host root.com
+	${UTIL} -make-intermediate -cert-in ${DCVECTORS_DIR}/root.crt -key-in ${DCVECTORS_DIR}/root.key -out ${DCVECTORS_DIR}/example.crt -key-out ${DCVECTORS_DIR}/example.key -host example.com
+	${UTIL} -make-intermediate -cert-in ${DCVECTORS_DIR}/root.crt -key-in ${DCVECTORS_DIR}/root.key -out ${DCVECTORS_DIR}/client_facing.crt -key-out ${DCVECTORS_DIR}/client_facing.key -host client-facing.com
+	${UTIL} -make-dcvectors -cert-in ${DCVECTORS_DIR}/example.crt -key-in ${DCVECTORS_DIR}/example.key -dc-algo 0x0807 -out ${DCVECTORS_DIR}/dc_ed25519.txt
+	${UTIL} -make-dcvectors -cert-in ${DCVECTORS_DIR}/example.crt -key-in ${DCVECTORS_DIR}/example.key -dc-algo 0x0403 -out ${DCVECTORS_DIR}/dc_ecdsa_p256.txt
+	${UTIL} -make-dcvectors -cert-in ${DCVECTORS_DIR}/example.crt -key-in ${DCVECTORS_DIR}/example.key -dc-algo 0x0503 -out ${DCVECTORS_DIR}/dc_ecdsa_p384.txt
+	${UTIL} -make-dcvectors -cert-in ${DCVECTORS_DIR}/example.crt -key-in ${DCVECTORS_DIR}/example.key -dc-algo 0x0603 -out ${DCVECTORS_DIR}/dc_ecdsa_p512.txt
```
Comment (cjpatton) [inline Makefile:39]: Is `-make-dcvectors` meant to replace `-make-dc`? It seems like it can, since the new option is more general.

**Review by cjpatton — COMMENTED** [review]
> To me, it seems like `-make-dcvectors` and `-make-dc` are the same thing. Also, instead of a `-dc-algo` argument, we could have an `-alg` argument that is passed to `make-dc`, `make-root`, etc. I say `alg` and not `sig-alg` because we might potentially wrap post-quantum KEMs in DCs.

I agree .... the new option is much more general. Though I would just stick with "-alg" or "-dc-algo" and not differentiate between KEM/signatures at this stage.

**Comment by claucece** [comment]
> To me, it seems like -make-dcvectors and -make-dc are the same thing. Also, instead of a -dc-algo argument, we could have an -alg argument that is passed to make-dc, make-root, etc. I say alg and not sig-alg because we might potentially wrap post-quantum KEMs in DCs.

I like that of the -alg argument. Around the ` -make-dc` one: I wanted to separate between the vectors and the dc used to run the test. But it could also be that the all of them are generated and the test just chooses one.

File: cmd/util/common.go (hunk 1 above)
Comment (claucece) [reply cmd/util/common.go:107]: Ah.. this wasn't the comment ;) I putted it there to remember ;) But I can add the comments as well.

File: Makefile (hunk 2 above)
Comment (claucece) [reply Makefile:39]: I'll unify them ;)

File: cmd/util/common.go (hunk 1 above)
Comment (cjpatton) [reply cmd/util/common.go:107]: Ha no worries. I guess I should hold off on these kinds of comments until it's no longer "WIP" and is "ready for review"?

**Comment by xvzcf** [comment]
> I like that of the -alg argument. Around the ` -make-dc` one: I wanted to separate between the vectors and the dc used to run the test. But it could also be that the all of them are generated and the test just chooses one.

Yeah, I was thinking the latter. For each testcase, the CI runner will dynamically generate the artifacts, run a client-server interaction and then clean up by deleting the artifacts. If (for example) signature algorithms for the root, intermediate cert, and DC are chosen randomly for each testcase run, this might get us decent coverage of the many possible combinations without too much additional complexity.

File: cmd/util/common.go (hunk 1 above)
Comment (claucece) [reply cmd/util/common.go]: Solved now. Open for review as well.

**Comment by claucece** [comment]
@xvzcf @cjpatton check how it is working now. It is not so simple to choose randomly a cert, as it has to be a valid cert for the dc that are generated upon them. For the moment, it is randomly choosing a DC.

File: cmd/util/sign.go [hunk 3]
```diff
@@ -12,92 +12,93 @@ package main
...
 	bugs  *CertificateBugs
 	curve elliptic.Curve
 	hash  crypto.Hash
-	priv  *ecdsa.PrivateKey
+	priv  crypto.PrivateKey
 	rand  io.Reader
+	ecdsa bool
 }
 
-func maybeCorruptECDSAValue(n *big.Int, typeOfCorruption BadValue, limit *big.Int) *big.Int {
```
Comment (xvzcf) [inline cmd/util/sign.go:32]: Why was this removed?

File: impl-endpoints/boringssl/run_endpoint.sh [hunk 4]
```diff
@@ -14,5 +14,6 @@ else
     echo "Running BoringSSL server."
     echo "Server params: $SERVER_PARAMS"
     echo "Test case: $TESTCASE"
-    bssl server -accept 4433 -cert /test-inputs/example.crt -key /test-inputs/example.key -subcert /test-inputs/dc.txt
+    DC=$(find /test-inputs/ -type f -name \*.txt | shuf -n 1)
```
Comment (xvzcf) [inline impl-endpoints/boringssl/run_endpoint.sh:17]: I would rather there be no selection of this sort in the endpoint, since it would have to be repeated everywhere, and, like you noted, it doesn't easily apply to certs.

What I was thinking was having a `make-root -alg 0` or `make-root -alg 0xFFFF` option (similarly for `make-intermediate` and `make-dc`), that would prompt the corresponding functions to randomly pick a signature algorithm from the list [here](https://github.com/xvzcf/tls-interop-runner/blob/dbb655f4a34cdab4ddfbf70b6611c78de593fb80/cmd/util/common.go#L34) and generate the artifact. The endpoints can then pick these up as usual.

File: cmd/util/common.go [hunk 5]
```diff
@@ -91,6 +91,7 @@ type Config struct {
 	// implementations.
 	Bugs CertificateBugs
 
+	// SignatureAlgorithm defines the signature algorithm for a Delegated Credential
```
Comment (xvzcf) [inline cmd/util/common.go:94]: I think this applies not just to DCs.

File: cmd/util/sign.go (hunk 3 above)
Comment (claucece) [reply cmd/util/sign.go:32]: I'm not really sure why you had it in the first place. In order to support different algorithms, you can use the crypto.Signer interface where this check does not make sense anymore, as the signature returned is already a byte array. Why was it needed if the signature is generated by the golang code?

Comment (claucece) [reply cmd/util/sign.go:32]: So, my question is more: why was the check there before?

File: impl-endpoints/boringssl/run_endpoint.sh (hunk 4 above)
Comment (claucece) [reply impl-endpoints/boringssl/run_endpoint.sh:17]: Right. But the point was having somewhere all of the generated DC as tests vectors, which means generating them all at the same time, rather than making them be randomly generated each one every time,

File: cmd/util/sign.go (hunk 3 above)
Comment (xvzcf) [reply cmd/util/sign.go:32]: Ah no, this wasn't a check. This was to malform the ECDSA signature, which we planned to use in future fuzzing tests. Perhaps it should be named better ...

Comment (claucece) [reply cmd/util/sign.go:32]: Ahhh hahaha.. now it makes sense ;) Ok, I can create a func that makes the byte array empty or changes some bytes, if corruption is needed.

File: impl-endpoints/boringssl/run_endpoint.sh (hunk 4 above)
Comment (claucece) [reply impl-endpoints/boringssl/run_endpoint.sh:17]: We talk this offline, and have an strategy going forward.

File: cmd/util/common.go (hunk 5 above)
Comment (claucece) [reply cmd/util/common.go:94]: Right. I'll change the comment.

File: cmd/util/sign.go (hunk 3 above)
Comment (xvzcf) [reply cmd/util/sign.go:32]: Yeah, we have the `Bugs CertificateBugs` struct to control this, and we hope to adding much more to that.

File: impl-endpoints/boringssl/run_endpoint.sh (hunk 4 above)
Comment (claucece) [reply impl-endpoints/boringssl/run_endpoint.sh]: Ok. I added a random way to choose in the Makefile itself. But this will be replaced later for golang generation itself.

File: cmd/util/sign.go (hunk 3 above)
Comment (claucece) [reply cmd/util/sign.go:32]: I added now some corruption, but it can be extended.

File: cmd/util/common.go (hunk 5 above)
Comment (claucece) [reply cmd/util/common.go]: Changed now.

**Comment by claucece** [comment]
This is done now. I added some todos regarding:

* Generating the randomness for the algorithms to be chosen should be done by golang itself as discussed with @xvzcf 
* Generating certificates with all other tls 1.3 supported algorithms need to be supported.

I can send that as part of other PRs.

Let me know what you think @xvzcf

File: cmd/util/util.go [hunk 6]
```diff
@@ -14,12 +14,13 @@ import (
 
 const usage = `Usage:
 
-    $ util [-help] {-make-root|-make-intermediate|-make-dc} [-cert-in PATH] [-key-in PATH] [-out PATH] [-key-out PATH]
+    $ util [-help] {-make-root|-make-intermediate|-make-dc|-make-dcvectors} [-cert-in PATH] [-key-in PATH] [-out PATH] [-key-out PATH]
```
Comment (xvzcf) [inline cmd/util/util.go:17]: I take it `make-dcvectors` should be removed?

File: cmd/util/sign.go [hunk 7]
```diff
@@ -12,92 +12,112 @@ package main
...
 	_ "crypto/sha512"
-	"encoding/asn1"
 	"fmt"
 	"io"
-	"math/big"
 )
 
-type ECDSASigner struct {
+// Signer represents an structure holding the signing information
+type Signer struct {
```
Comment (xvzcf) [inline cmd/util/sign.go:24]: This should be fine for now, but I'm guessing once we handle RSA we should probably have a separate struct for that.

Comment (claucece) [reply cmd/util/sign.go:24]: It could also be that we have a bool for each case: isRSA, is ed25519 and `curve` param for eddsa.

File: cmd/util/util.go (hunk 6 above)
Comment (claucece) [reply cmd/util/util.go:17]: Oh, yes! A forgotten one here ;)

**Review by xvzcf — APPROVED** [review]
As discussed, the TODOs can be addressed once I break out the common code into an `internal` library.
